protoc-gen-openapiv2/options/openapiv2.proto

```diff
@@ -505,6 +505,11 @@ message JSONSchema {
   reserved 43 to 45;
   // Items in `enum` must be unique https://tools.ietf.org/html/draft-fge-json-schema-validation-00#section-5.5.1
   repeated string enum = 46;
+  // Alternative parameter name when used as path parameter. If set, this will
+  // be used as the complete parameter name when this field is used as a path
+  // parameter. Use this to avoid having auto generated path parameter names
+  // for overlapping paths.
+  string path_param_name = 47;
```

Comment:
I appreciate the initiative in opening this PR, but I think, so far, we've only added fields here for valid fields of the equivalent OpenAPIv2 spec object, which in this case is the `Schema` object: https://github.com/OAI/OpenAPI-Specification/blob/3.0.0/versions/2.0.md#schemaObject. Adding this field breaks this pattern. Is there some alternative way we can inject this information?

Reply:
That's a fair point, and the same thought struck me as well when implementing this. How about introducing separate options for configuring the behaviour of the generator keep it separated from the OpenAPI objects? It looks like that requires a separate assigned message id in `annotations.proto` though

Reply:
So now that I've eaten my own dog food regarding the path parameters, the automatic name generation kinda sucks (just as outlined above) especially when you start to have a lot of API endpoints that all have a similar structure.  I've also discovered a bug (though I don't have reproduction steps) where you can get duplicate endpoints generated for the same actual API endpoint. (I'll figure out a simple repro so I can file it at some point).

I really love the idea of being able to name the path parameters but i also understand @johanbrandhorst point that the openapiv2 proto is really meant for openapi2 specific attributes.

I'd really love it if we could have some way of doing this.  I'm not sure what the best way is though.  Adding and additional extension just for this one concept is a pretty big ask as the field number would have to be assigned by the protobuf-global-extension-registry.

But adding a field to the JSONSchema message also seems wrong.

Maybe there could be a special syntax of the google.api.http attribute that specifies the name as it would be in the URL like:
```
option(google.api.http) = {
   get: "/v1/{name=foo/*:fooName}"
}
```

Reply:
I don't want to go down the path of having the openapiv2 generator mess with the http.proto annotations either (they are separate today). Is this operation something that can be done as a post-processing step of the swagger file? Maybe the best option here is to provide some tooling for users to perform post-processing of names... though that would require keeping a mapping somewhere.

Reply:
Good point about post processing, but it won't work well, as the operations get mixed between paths. It's not just a matter of replacing `{name}` and `{name_1}`. Both the path and the parameter name must be replaced as well.

IMO, `JSONSchema` is specific to protoc-gen-openapiv2 (just look at all the reserved fields), and could contain fields that doesn't correspond to the OpenAPI spec. We could jump field numbers to 1000 for the fields related to transformation.

Note that extension 1042 is reserved for "grpc-gateway protoc-gen-swagger", so it's this project that decides how it's used: https://chromium.googlesource.com/chromium/src/+/refs/heads/main/third_party/protobuf/docs/options.md

Reply:
How about something like this? This pattern could be applied to other elements as well if needed. I guess it's crucial that the chosen name (e.g. `field_configuration`) is unlikely to clash with future  names in OpenAPI. Maybe a name like `proto_configuration` or `proto_transformation` is safer.

```
message JSONSchema {
  …
  FieldConfiguration field_configuration = 1000;
}

// 'FieldConfiguration' provides additional field level properties used when generating the OpenAPI v2 file.
message FieldConfiguration {
  // Alternative parameter name when used as path parameter. If set, this will
  // be used as the complete parameter name when this field is used as a path
  // parameter. Use this to avoid having auto generated path parameter names
  // for overlapping paths.
  string path_param_name = 47;
}
```

Reply:
I think this could work. I'm not too worried about new fields being added to the OpenAPIv2 spec definition of JSONSchema since that version is pretty much deprecated already. I would welcome @ivucica's thoughts on this as well.

Reply:
Been lurking around this change for a bit and have been dealing with the path parameter name problem for quite some time so I'll chime in here. 

RE: post processing, that's where i've found the most success. As @yinzara commented the name generation is still kinda meh. If you're using this generated spec as an input to additional tooling like postman losing out on the path structure kinda sucks.

Even when we go from `parent=books/*` > `/v1/{bookParent}` we lose out on quite a bit of context.
Retaining the path is potentially more desirable `parent=books/*` > `/v1/books/{book_id}`.

This is further exacerbated when you have even more complex relationships like. `name=projects/*/shelves/*/books/*` > `/v1/{bookName}` vs `/v1/projects/{project_id}/shelves/{shelf_id}/books/{book_id}`

Here's a gist of a some small script i've used to post process these files to achieve some of what i've needed. It's not perfect or a complete solution but felt it may offer some insight into the viability to a solution that relies on post processing.

https://gist.github.com/dgparker/80d3fc815bfa21c98cba70d46ed1f7ab

Reply:
> Here's a gist of a some small script i've used to post process these files to achieve some of what i've needed. It's not perfect or a complete solution but felt it may offer some insight into the viability to a solution that relies on post processing.

Interesting - thanks for sharing.

I am correct that you run this on the output from version 2.7.1, given this regex `(?P<start>{.*=)(?P<binding>[\w\/\*]+)(?P<end>})`?

Reply:
@dgparker @oyvindwe if this is sufficient for your needs, I'd prefer not to make the change here, obviously. It would be great to have a little guide for this sort of thing written up for our docs. Would you be willing to contribute something? Thanks!

Reply:
I think that my change to make the specification valid makes this solution impossible with any version of grpc-gateway beyond 2.7.1

Reply:
I think that my change to make the specification valid makes this solution impossible with any version of grpc-gateway beyond 2.7.1

Reply:
> I think that my change to make the specification valid makes this solution impossible with any version of grpc-gateway beyond 2.7.1

I agree - the post processing depends on an invalid OpenAPI file, which is kind of the wrong starting point. The context needed to do the post processing is lost in the valid file. It could be possible to rewrite the post processing to use the `pattern` field of path parameters though.

If going down this route, we could as well include the same logic in the generator, and enable it with a CLI argument. I'm still not entirely happy about it, as there is no way to provide parameter descriptions and examples.

Reply:
I just realised that supporting paths transforms like `{name=shelves/*}` -> `shelves/{shelf_id}` requires a stemming dictionary to find the singular (to avoid results like `shelves/{shelve_id}`). This makes me even more uncertain of this approach.

Reply:
I'm slightly lost, are we suggesting that neither post processing nor annotations are going to work?

Reply:
@johanbrandhorst apologies for late follow up from my side.

Regarding post processing as suggested by @dgparker : This no longer works due to @yinzara 's path fixes. Also, as discussed in issue #720 this approach has several shortcomings; the generated path parameters lacks descriptions, type information, and examples values. Also, as pointed out above, creating singular from plural isn't straight forward in English. I suggest we continue the discussion on how to generate path parameters from a path template in that issue (e.g `{name=shelves/*/book/*}` -> `shelves/{shelf_id}/books/{book_id}`).

This PR suggests an annotation to override the generated path parameter (name, name_1, name_2, etc.). This path parameter would use the same description, type, and example value as currently used. The only issue raised with this PR is that it will introduce a field to `JSONSchema` that isn't a part of the OpenAPI v2 specification. Technically, this PR works fine.

Reply:
Thanks for the update. I see we need a rebase/merge of this PR, but I think unless there are strong objects from others that adding the annotation might be okay. This is a very real request from users and I'm keen to allow this to be configured.